package/scripts/prmon_compress_output.py

```diff
@@ -0,0 +1,159 @@
+    metric_series = df[metric]
+    metric_redux = metric_series.copy()
+    dyn_range = metric_series.max() - metric_series.min()
+    eps = dyn_range * precision
+    idx = 0
+    while True:
+        metriclen = len(metric_redux)
+        if idx == metriclen - 2:
+            break
+        p1 = (metric_redux.index[idx], metric_redux.iloc[idx])
+        p2 = (metric_redux.index[idx + 1], metric_redux.iloc[idx + 1])
+        p3 = (metric_redux.index[idx + 2], metric_redux.iloc[idx + 2])
+        if interp_drop(p1, p2, p3, eps):
+            metric_redux = metric_redux.drop(metric_redux.index[idx + 1])
+        else:
+            idx += 1
+    return metric_redux
+
+
+def reduce_steady_metric(df, metric):
+    """For more steady metrics just keep the changing points"""
+    metric = df[metric]
+    return metric[metric != metric.shift(1)]
+
+
+def compress_prmon_output(df, precision, skip_interpolate):
+    """Compress full df. Final index is the union of the compressed series indexes.
+    Points without values for a series are either linearly interpolated,
+    for fast-changing metrics, or forward-filled, for steady metrics"""
+    if len(df) > 2:
+        present_changing_metrics = [
+            metric for metric in CHANGING_METRICS if metric in df.columns
+        ]
+        present_steady_metrics = [
+            metric for metric in STEADY_METRICS if metric in df.columns
+        ]
+        reduced_changing_metrics = [
+            reduce_changing_metric(df, metric, precision)
+            for metric in present_changing_metrics
+        ]
+        reduced_steady_metrics = [
+            reduce_steady_metric(df, metric) for metric in present_steady_metrics
+        ]
+        final_df = pd.concat(reduced_changing_metrics + reduced_steady_metrics, axis=1)
+        if not skip_interpolate:
+            final_df[present_changing_metrics] = final_df[
+                present_changing_metrics
+            ].interpolate(method="index")
+            final_df[present_steady_metrics] = final_df[present_steady_metrics].ffill(
+                downcast="infer"
+            )
+            final_df = final_df.round(0)
+        final_df = final_df.astype("Int64", errors="ignore")
+        return final_df
+    return df
+
+
+def main():
+    """Main compression function"""
+    parser = argparse.ArgumentParser(
+        description="Configurable smart compression script"
+    )
+
+    parser.add_argument(
+        "--input",
+        type=str,
+        default="prmon.txt",
+        help="PrMon TXT output that will be used as input",
+    )
+
+    parser.add_argument(
+        "--output",
+        type=str,
+        default="prmon_compressed.txt",
+        help="name of the output compressed text file",
+    )
+
+    parser.add_argument(
+        "--precision",
+        type=lambda x: float(x)
+        if 0 < float(x) < 1
+        else parser.exit(-1, "Precision must be strictly between 0 and 1"),
+        default=0.1,
```

Comment:
I believe 0.05 is a better default, at least going over the example we have here:

**0.1 drops the file size from 50 KB to 15 KB, w/:**
![reduced-new-default](https://user-images.githubusercontent.com/7428938/106144338-5976f180-6174-11eb-9a47-f5967c0a1e96.png)

**0.05 drops the file size from 50 KB to 16 KB, w/:**
![reduced-new-0 05](https://user-images.githubusercontent.com/7428938/106144336-5845c480-6174-11eb-8ef9-15afed69c63b.png)

Could you check and confirm? It looks like a price worth paying for 😄

Reply:
This was a very open point, and it still is.

We also had some issues with deleting too much information on some test cases so I am happy to change the default.